python/pyspark/ml/tuning.py

```diff
@@ -255,18 +257,23 @@ def _fit(self, dataset):
         randCol = self.uid + "_rand"
         df = dataset.select("*", rand(seed).alias(randCol))
         metrics = [0.0] * numModels
+
+        pool = ThreadPool(processes=min(self.getParallelism(), numModels))
+
         for i in range(nFolds):
             validateLB = i * h
             validateUB = (i + 1) * h
             condition = (df[randCol] >= validateLB) & (df[randCol] < validateUB)
-            validation = df.filter(condition)
+            validation = df.filter(condition).cache()
```

Comment:
Here maybe need a discussion.
Currently in pyspark it both do not cache `train dataset` and `validation dataset` but in scala impl it cache both of them.
But I prefer cache `validation dataset` but do not cache `train dataset`, because the size of `validation dataset` is only `1/numFolds` of input dataset, it deserve caching otherwise it will scan input dataset again. But the size `train dataset` is `(numFolds - 1)/numFolds` of input dataset. We can directly scan from input dataset to generate the `train dataset` and won't slow down too much.
@BryanCutler @MLnick What do you think about it ? Thanks!

Reply:
We will do multi-model training when fitting on the estimator. So I think it's still beneficial to cache training dataset?

Reply:
Suppose we have already cached input dataset, then generate "training dataset" only need a "map" operation on cached df with filtering out only `1/numFolds`. So I think the cost won't be too much more compared with caching "training dataset".

Reply:
That's right, but seems we don't check if input dataset is cached or not here? Should we cache it if it is not cached?

Reply:
Hmm... How to checking input dataset caching status is not easy, there are still discussions in `SPARK-18608`. But for now I think we can keep consistent with scala-side to cache both of `training` and `validation` dataset. So I update code.